src/vs/workbench/services/search/node/textSearch.ts

```diff
@@ -5,330 +5,178 @@
+			const worker = this.workers[this.nextWorker];
+			this.nextWorker = (this.nextWorker + 1) % this.workers.length;
 
-			// Indicate progress to the outside
-			progress();
-
-			let fileMatch: FileMatch = null;
-
-			let doneCallback = (error?: Error) => {
-				if (!error && !this.isCanceled && fileMatch && !fileMatch.isEmpty()) {
-					onResult(fileMatch.serialize());
-				}
-
-				return unwind(size);
-			};
-
-			const absolutePath = result.base ? [result.base, result.relativePath].join(path.sep) : result.relativePath;
-			let perLineCallback = (line: string, lineNumber: number) => {
-				if (this.limitReached || this.isCanceled) {
-					return; // return early if canceled or limit reached
+			const maxResults = this.config.maxResults - this.numResults;
+			worker.search({ absolutePaths: batch, maxResults }).then(result => {
+				if (!result || this.limitReached || this.isCanceled) {
+					return unwind(batchBytes);
 				}
 
-				let lineMatch: LineMatch = null;
-				let match = this.contentPattern.exec(line);
-
-				// Record all matches into file result
-				while (match !== null && match[0].length > 0 && !this.limitReached && !this.isCanceled) {
-					resultCounter++;
-					if (this.maxResults && resultCounter >= this.maxResults) {
-						this.limitReached = true;
-					}
-
-					if (fileMatch === null) {
-						fileMatch = new FileMatch(absolutePath);
-					}
-
-					if (lineMatch === null) {
-						lineMatch = new LineMatch(line, lineNumber);
-						fileMatch.addMatch(lineMatch);
-					}
-
-					lineMatch.addMatch(match.index, match[0].length);
+				const matches = result.matches;
+				this.numResults += result.numMatches;
+				matches.forEach(m => {
+					onResult(m);
+				});
 
-					match = this.contentPattern.exec(line);
+				if (this.numResults >= this.config.maxResults) {
+					// It's possible to go over maxResults like this, but it's much simpler than trying to extract the exact number
+					// of file matches, line matches, and matches within a line to == maxResults.
+					this.limitReached = true;
 				}
-			};
 
-			// Read lines buffered to support large files
-			this.readlinesAsync(absolutePath, perLineCallback, { bufferLength: 8096, encoding: this.fileEncoding }, doneCallback);
-		}, (error, isLimitHit) => {
-			this.walkerIsDone = true;
-			this.walkerError = error;
-			unwind(0 /* walker is done, indicate this back to our handler to be able to unwind */);
-		});
-	}
+				unwind(batchBytes);
+			});
+		};
 
-	private readlinesAsync(filename: string, perLineCallback: (line: string, lineNumber: number) => void, options: ReadLinesOptions, callback: (error: Error) => void): void {
-		fs.open(filename, 'r', null, (error: Error, fd: number) => {
-			if (error) {
-				return callback(error);
+		// Walk over the file system
+		let nextBatch = [];
+		let nextBatchBytes = 0;
+		const batchFlushBytes = 2 ** 20; // 1MB
+		this.walker.walk(this.config.rootFolders, this.config.extraFiles, result => {
+			let bytes = result.size || 1;
```

Comment:
Maybe so we continue to report progress while searching empty files. It won't break that check though.